### PR TITLE
mkfs-util: Escape spaces with slashes in protofile format

### PR DESCRIPTION
**Pull request by daandemeyer**

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
See https://lore.kernel.org/linux-xfs/20230222090303.h6tujm7y32gjhgal@andromeda/T/#m8066b3e7d62a080ee7434faac4861d944e64493b

File: src/shared/mkfs-util.c [hunk 1]
```diff
@@ -204,15 +204,29 @@ static int protofile_print_item(
...
+         * for more information.*/
+
+        if (S_ISREG(sx->stx_mode)) {
+                _cleanup_free_ char *p = NULL;
+
+                p = strdup(path);
+                if (!p)
+                        return log_oom();
+
+                string_replace_char(p, ' ', '/');
```
Comment (poettering) [inline src/shared/mkfs-util.c]: how frickin frickin frickin ugly. Christ.

**Comment by poettering** [comment]
lgtm, fwitw.

i guess you are waiting until this is merged upstream?

**Comment by DaanDeMeyer** [comment]
Yes the API is still being discussed, they might hide it behind an option which would make it a PITA to consume as we'd have to do version checks on mkfs to see if the option is available.

**Comment by DaanDeMeyer** [comment]
So the API changed slightly and the option is now behind a flag `slashes_are_spaces`. I modified the PR to set that flag if we encounter any filenames with spaces. The only annoyance is that older mkfs.xfs versions will fail with "mkfs.xfs: failed to open slashes_are_spaces=1,/home/daandemeyer/projects/systemd/.mkosi.tmpuzagg859/.#5639e5fa53655fd1: No such file or directory" if there are filenames with spaces. On the other hand, we currently also fail with a non-descriptive error if we encounter filenames with spaces, so it's not like the situation is worse with this PR.

File: man/repart.d.xml [hunk 2]
```diff
@@ -433,7 +433,8 @@
 
         <para>Note that when populating XFS filesystems with <command>systemd-repart</command> and loop
         devices are not available, populating XFS filesystems with files containing spaces, tabs or newlines
-        will fail due to limitations of mkfs.xfs's protofile format.</para>
+        might fail on old versions of <command>mkfs.xfs</command> due to limitations of mkfs.xfs's protofile
```
Comment (poettering) [inline man/repart.d.xml]: enclose both `mkfs.xfs` in `<command>`?

**Comment by poettering** [comment]
lgtm. I guess you want to drop the "draft" state now?

**Comment by poettering** [comment]
lgtm, feel free to set to green label once draft mode is dropped

**Comment by DaanDeMeyer** [comment]
Will merge this once a new version of xfsprogs with the flag in it is released.

File: man/repart.d.xml [hunk 3]
```diff
@@ -435,9 +435,9 @@
 
         <para>Note that when populating XFS filesystems with <command>systemd-repart</command> and loop
         devices are not available, populating XFS filesystems with files containing spaces, tabs or newlines
-        will fail due to limitations of <citerefentry
-        project='man-pages'><refentrytitle>mkfs.xfs</refentrytitle><manvolnum>8</manvolnum></citerefentry>
-        protofile format.</para>
+        might fail on old versions on old versions of
```
Comment (bluca) [inline man/repart.d.xml]: "on old versions on old versions"

**Review by bluca — APPROVED** [review]

**Comment by DaanDeMeyer** [comment]
Changed to always use /tmp for the symlink directory instead of TMPDIR if set because we want to be 100% sure that there's no spaces in the directory name.

**Pull request merged**